### PR TITLE
Shv3 MetaMethods

### PR DESCRIPTION
**Pull request by syyyr**

### CODE REVIEW

File: libshvchainpack/src/chainpack/metamethod.cpp [hunk 1]
```diff
@@ -50,68 +72,47 @@ const std::string& MetaMethod::description() const
...
 }
 
-MetaMethod& MetaMethod::setTag(const std::string &key, const RpcValue& value)
-{
-	m_tags.setValue(key, value); return *this;
-}
+enum class Ikey {
+	Name = 1,
+	Flags = 2,
+	Param = 3,
```
Comment (fvacek) [inline libshvchainpack/src/chainpack/metamethod.cpp:93]: ParamType

File: libshvchainpack/src/chainpack/metamethod.cpp [hunk 2]
```diff
@@ -50,68 +72,47 @@ const std::string& MetaMethod::description() const
...
+	Result = 4,
```
Comment (fvacek) [inline libshvchainpack/src/chainpack/metamethod.cpp:94]: ResultType

File: libshvchainpack/include/shv/chainpack/metamethod.h [hunk 3]
```diff
@@ -42,22 +42,34 @@ class SHVCHAINPACK_DECL_EXPORT MetaMethod
 	static constexpr auto KEY_LABEL = "label";
 	static constexpr auto KEY_TAGS = "tags";
 public:
+	struct Signal {
+		Signal(std::string name, std::string result);
+		Signal(std::string name);
+		std::string name;
+		RpcValue result;
```
Comment (fvacek) [inline libshvchainpack/include/shv/chainpack/metamethod.h:49]:
```suggestion
		std::string paramType;
```

File: libshviotqt/src/node/filenode.cpp [hunk 4]
```diff
@@ -22,21 +22,22 @@ const auto M_READ_COMPRESSED = "readCompressed";
...
-	{cp::Rpc::METH_DIR, cp::MetaMethod::Signature::RetParam, cp::MetaMethod::Flag::None, cp::Rpc::ROLE_BROWSE},
-	{cp::Rpc::METH_LS, cp::MetaMethod::Signature::RetParam, cp::MetaMethod::Flag::None, cp::Rpc::ROLE_BROWSE},
-	{M_HASH, cp::MetaMethod::Signature::RetVoid, cp::MetaMethod::Flag::None, cp::Rpc::ROLE_READ},
-	{M_SIZE, cp::MetaMethod::Signature::RetVoid, cp::MetaMethod::Flag::LargeResultHint, cp::Rpc::ROLE_BROWSE},
-	{M_SIZE_COMPRESSED, cp::MetaMethod::Signature::RetParam, cp::MetaMethod::Flag::None, cp::Rpc::ROLE_BROWSE
+	// FIXME: Add proper types param and ret types
+	{cp::Rpc::METH_DIR, cp::MetaMethod::Flag::None, "param", "ret", cp::MetaMethod::AccessLevel::Browse},
+	{cp::Rpc::METH_LS, cp::MetaMethod::Flag::None, "param", "ret", cp::MetaMethod::AccessLevel::Browse},
+	{M_HASH, cp::MetaMethod::Flag::None, "ret", "void", cp::MetaMethod::AccessLevel::Read},
+	{M_SIZE, cp::MetaMethod::Flag::LargeResultHint, "void", "ret", cp::MetaMethod::AccessLevel::Browse},
```
Comment (fvacek) [inline libshviotqt/src/node/filenode.cpp:29]:
```suggestion
	{M_SIZE, cp::MetaMethod::Flag::LargeResultHint, "UInt", "void", cp::MetaMethod::AccessLevel::Browse},
```

File: libshviotqt/src/node/filenode.cpp [hunk 5]
```diff
@@ -22,21 +22,22 @@ const auto M_READ_COMPRESSED = "readCompressed";
...
+	{M_SIZE_COMPRESSED, cp::MetaMethod::Flag::None, "ret", "param", cp::MetaMethod::AccessLevel::Browse , {}
 	 , "Parameters\n"
 	   "  read() parameters\n"
 	   "  compressionType: gzip (default) | qcompress"
 	},
-	{M_READ, cp::MetaMethod::Signature::RetVoid, cp::MetaMethod::Flag::LargeResultHint, cp::Rpc::ROLE_READ
+	{M_READ, cp::MetaMethod::Flag::LargeResultHint, "void", "ret", cp::MetaMethod::AccessLevel::Read, {}
```
Comment (fvacek) [inline libshviotqt/src/node/filenode.cpp:35]:
```suggestion
	{M_READ, cp::MetaMethod::Flag::LargeResultHint, "void", "Blob", cp::MetaMethod::AccessLevel::Read, {}
```

File: libshvchainpack/include/shv/chainpack/metamethod.h [hunk 6]
```diff
@@ -69,16 +81,28 @@ class SHVCHAINPACK_DECL_EXPORT MetaMethod
...
-	static const char* accessLevelToString(int access_level);
+	static const char* accessLevelToString(AccessLevel access_level);
 private:
 	std::string m_name;
-	Signature m_signature = Signature::VoidVoid;
 	unsigned m_flags = 0;
-	RpcValue m_accessGrant;
+	std::string m_param;
+	std::string m_result;
+	AccessLevel m_accessGrant;
```
Comment (fvacek) [inline libshvchainpack/include/shv/chainpack/metamethod.h:90]:
```suggestion
	AccessLevel m_accessLevel;
```

File: libshvchainpack/include/shv/chainpack/metamethod.h [hunk 7]
```diff
@@ -42,22 +42,34 @@ class SHVCHAINPACK_DECL_EXPORT MetaMethod
...
 	bool isValid() const;
 	const std::string& name() const;
 	const std::string& label() const;
 	MetaMethod& setLabel(const std::string &label);
 	const std::string& description() const;
-	Signature signature() const;
 	unsigned flags() const;
-	const RpcValue& accessGrant() const;
-	const RpcValue::Map& tags() const;
+	RpcValue accessGrant() const;
```
Comment (fvacek) [inline libshvchainpack/include/shv/chainpack/metamethod.h:71]:
```suggestion
	AccessLevel accessLevel() const;
```

**Review by fvacek — CHANGES_REQUESTED** [review]

File: libshvchainpack/include/shv/chainpack/metamethod.h (hunk 6 above)
Comment (syyyr) [reply libshvchainpack/include/shv/chainpack/metamethod.h]: done

File: libshvchainpack/src/chainpack/metamethod.cpp (hunk 2 above)
Comment (syyyr) [reply libshvchainpack/src/chainpack/metamethod.cpp]: done

File: libshvchainpack/src/chainpack/metamethod.cpp (hunk 1 above)
Comment (syyyr) [reply libshvchainpack/src/chainpack/metamethod.cpp]: done

File: libshvchainpack/include/shv/chainpack/metamethod.h [hunk 8]
```diff
@@ -42,22 +42,34 @@ class SHVCHAINPACK_DECL_EXPORT MetaMethod
 	static constexpr auto KEY_LABEL = "label";
 	static constexpr auto KEY_TAGS = "tags";
 public:
+	struct Signal {
+		Signal(std::string name, std::string result);
```
[the remark this answers is not in the document]
Comment (syyyr) [reply libshvchainpack/include/shv/chainpack/metamethod.h]: done

File: libshvchainpack/include/shv/chainpack/accessgrant.h [hunk 9]
```diff
@@ -61,7 +61,7 @@ struct SHVCHAINPACK_DECL_EXPORT AccessGrant
```
[the remark this answers is not in the document]
Comment (syyyr) [reply libshvchainpack/include/shv/chainpack/accessgrant.h]: done

File: libshviotqt/src/node/filenode.cpp (hunk 4 above)
Comment (syyyr) [reply libshviotqt/src/node/filenode.cpp:29]: done

File: libshviotqt/src/node/filenode.cpp (hunk 5 above)
Comment (syyyr) [reply libshviotqt/src/node/filenode.cpp:35]: Done

File: libshvchainpack/include/shv/chainpack/metamethod.h (hunk 7 above)
Comment (syyyr) [reply libshvchainpack/include/shv/chainpack/metamethod.h:71]: Done

File: examples/cli/minimalshvbroker/src/application.cpp [hunk 10]
```diff
@@ -20,8 +20,8 @@ class TestNode : public shv::iotqt::node::MethodsTableNode
 	TestNode(shv::iotqt::node::ShvNode *parent = nullptr)
 		: Super(std::string(), &m_metaMethods, parent)
 		, m_metaMethods{
-			{cp::Rpc::METH_DIR, cp::MetaMethod::Signature::RetParam, cp::MetaMethod::Flag::None, cp::MetaMethod::AccessLevel::Browse},
-			{cp::Rpc::METH_LS, cp::MetaMethod::Signature::RetParam, cp::MetaMethod::Flag::None, cp::MetaMethod::AccessLevel::Browse},
+			{shv::chainpack::Rpc::METH_DIR, shv::chainpack::MetaMethod::Flag::None, "param", "ret", shv::chainpack::MetaMethod::AccessLevel::Browse},
```
[the remark this answers is not in the document]
Comment (syyyr) [reply examples/cli/minimalshvbroker/src/application.cpp]: done

**Pull request merged**